magenta-lib/src/main/scala/magenta/tasks/gcp/GCSTasks.scala

```diff
@@ -0,0 +1,130 @@
+import magenta.{DeploymentResources, KeyRing, Region, Stack, Stage}
+import magenta.artifact.{S3Location, S3Object, S3Path}
+import magenta.deployment_type.param_reads.PatternValue
+import magenta.tasks.Task
+import software.amazon.awssdk.core.internal.util.Mimetype
+import software.amazon.awssdk.services.s3.S3Client
+import software.amazon.awssdk.services.s3.model.GetObjectRequest
+
+case class GCSUpload(
+  region: Region,
+  bucket: String,
+  paths: Seq[(S3Location, String)],
+  cacheControlPatterns: List[PatternValue] = Nil,
+  extensionToMimeType: Map[String,String] = Map.empty,
+  publicReadAcl: Boolean = false,
+  detailedLoggingThreshold: Int = 10,
+)(implicit val keyRing: KeyRing, artifactClient: S3Client,
+  withClientFactory: (KeyRing, DeploymentResources) => (Storage => Unit) => Unit = GCS.withGCSClient[Unit]) extends Task with Loggable {
+
+  private val PublicAcl = Arrays.asList(new ObjectAccessControl().setEntity("allUsers").setRole("READER"))
+
+  lazy val objectMappings: Seq[(S3Object, GCSPath)] = paths flatMap {
+    case (file, targetKey) => resolveMappings(file, targetKey, bucket)
+  }
+
+  lazy val totalSize: Long = objectMappings.map{ case (source, _) => source.size }.sum
+
+  lazy val transfers: Seq[StorageObjectTransfer] = objectMappings.map { case (source, target) =>
+    val storageObject = new StorageObject()
+    storageObject.setBucket(bucket)
+    storageObject.setName(target.key)
+    storageObject.setSize(BigInteger.valueOf(source.size))
+    storageObject.setContentType(URLConnection.guessContentTypeFromName(target.key))
+
+    cacheControlLookup(target.key) match {
+      case Some(cacheControl) => storageObject.setCacheControl(cacheControl)
+      case None               => ()
+    }
+
+    if (publicReadAcl) storageObject.setAcl(PublicAcl)
+
+    StorageObjectTransfer(source, storageObject)
+  }
+
+  def fileString(quantity: Int) = s"$quantity file${if (quantity != 1) "s" else ""}"
+
+  // end-user friendly description of this task
+  def description: String = s"Upload ${fileString(objectMappings.size)} to GCS bucket $bucket using file mapping $paths"
+
+  // execute this task (should throw on failure)
+  override def execute(resources: DeploymentResources, stopFlag: => Boolean) {
+    if (totalSize == 0) {
+      val locationDescription = paths.map {
+        case (path: S3Path, _) => path.show()
+        case (location, _) => location.toString
+      }.mkString("\n")
+      resources.reporter.fail(s"No files found to upload in $locationDescription")
+    }
+
+    val withClient = withClientFactory(keyRing, resources)
+    withClient { client =>
+
+      resources.reporter.verbose(s"Starting transfer of ${fileString(objectMappings.size)} ($totalSize bytes)")
+      transfers.zipWithIndex.par.foreach { case (transfer, index) =>
+        logger.debug(s"Transferring $transfer")
+        index match {
+          case x if x < 10 => resources.reporter.verbose(s"Transferring $transfer")
```

Comment:
the `detailedLoggingThreshold` is intended to be used in place of the `10` here.